uips/uip-4.md

```diff
@@ -0,0 +1,102 @@
+# UIP: Spend Backreferences
+
+| uip | 4 |
+| - | - |
+| title | Spend Backreferences |
+| description | Spend Backreferences enable improved sync performance |
+| author | Jennifer Helsby ([@redshiftzero](https://github.com/redshiftzero)), Henry de Valence ([@hdevalence](https://github.com/hdevalence)), Lúcás Meier ([@cronokirby](https://github.com/cronokirby))|
+| discussions-to | <https://forum.penumbra.zone/t/uip-spend-backreferences/110> |
+| status | Draft |
+| type | Standards Track |
+| consensus | Yes |
+| created | 2024-11-06 |
+
+## Abstract
+
+This specification introduces a method to improve Penumbra sync speeds by adding additional data that can be used by DAGSync clients. `Spend` actions will contain a new `backref_commitment` field, allowing clients to traverse their transaction graph backwards and quickly recover their entire transaction history.
+
+## Motivation
+
+DAGSync is a graph-aware fast syncing algorithm. A client, upon detecting a single transaction involving them, can check that outputs visible to them are spent or not. If the output is unspent, then they have identified a live note they can potentially spend in the future, else if the output is unspent, they can continue the process forwards in the transaction graph, until they reach unspent notes.
+
+The design of Penumbra currently does not allow traversal backwards through the transaction graph, only forwards. A `Spend` intentionally does not reveal the note being spent, only the nullifier that is revealed. By including on the `Spend` an encrypted reference back to the note commitment being spent, such that only the note owner can view it, we enable DAGSync clients to efficienctly reconstruct the transaction history both backwards and forwards.
```

Comment:
I think there is a conceptual issue here that's worth discussing at some point in the UIP, namely, that note commitments uniquely identify **note contents**, but not **note instances**. Two identical notes can be created and added to the chain, and they will have the same note commitment, although they are different instances. (This is why nullifier derivation hashes the position of the note, and it's the same conceptual mismatch that creates the faerie gold attack).

In practice, note commitments are _usually_ also commitments to note instances, because honest users should generate random `rseed` values, which causes the commitments to be unique. But this isn't guaranteed by the protocol, and we should pay attention to the implications.

My suspicion is that the note commitment is still the right thing to put in the backreference, but it's worth thinking through the implications and writing that up in the UIP. What are the consequences of non-unique backreferences? Should the backreference also include the position, so that it is unique? (I believe the position is supplied as part of the `SpendPlan` data to be able to derive a nullifier, so I think this wouldn't introduce any new data/plumbing).

For instance, is a tagging attack possible along the following lines:

- User A sends user B a sequence of notes with identical note commitments
  - Maybe this happens over time so that user B spends each of them in turn (rather than having new ones displace old ones)
- User B resyncs, queries for transaction IDs by note commitment
  - Something happens? Or not? Maybe the RPC returns all matching TxIds, maybe it just returns one of them and some data is missing?
  
After thinking about it a little (the time it took to write this comment) I don't think there's a real issue here, but probably good to cover in the writeup.

Reply:
In terms of the graph, you might already have multiple backreferences in a given transaction, so if one backreference points to two transactions, or you have two backreferences pointing to two transactions, the syncing algorithm works the same way, so I don't think there's really an issue there.

I think we just need to make sure to be aware of the non-uniqueness, and not do some kind of clobbering accidentally.

Reply:
For the transaction ID by note commitment, I think the RPC should ideally be returning all matching TxIds, instead of the most recent one such that no data is "lost" - i.e. the Penumbra protocol already allows there to be duplicate note commitments and we should keep that in mind. User B would continue to DAGSync with the list of relevant TxIds they got for the entire transaction, and from there can recover their transaction history, so I think this is fine. It's definitely worth flagging this possibility in the UIP so I'll add some notes to the Rationale.

Reply:
One implication is that an attacker could attempt to spam a target user with notes that have duplicate note commitments such that DAGSyncing takes slightly longer. Fees should mitigate that though.

Reply:
added e04f4885c2be81cb3f6d8b09a4812f72b3adeb70